docs: Trace correlations (#104309)
grafana/grafana: docs/sources/datasources/tempo/traces-in-grafana/trace-correlations.md

  - grafana
  - tempo
  - guide
  - tracing
  - correlations
labels:
  products:
    - cloud
    - enterprise
    - oss
menuTitle: Trace correlations
title: Trace correlations
weight: 1000
---

# Trace correlations

You can use Grafana [correlations](/docs/grafana/<GRAFANA_VERSION>/administration/correlations/) to embed interactive correlation links in your trace view to jump from spans to related logs, metrics, profiles, or external systems. This guide explains how to configure and manage Trace correlations in Grafana.

## What are trace correlations?

Trace correlations let you define rules that inject context-sensitive links into your trace spans. When viewing traces in Explore or the Traces panel, users can click these links to navigate directly to relevant queries or URLs. Correlations are similar but more flexible to the [trace to logs, metrics, and profiles links you can configure for the Tempo data source](/docs/grafana/<GRAFANA_VERSION>/datasources/tempo/configure-tempo-data-source).

{{< figure src="/media/docs/tempo/screenshot-trace-view-correlations.png" max-width="900px" class="docs-image--no-shadow" alt="Using correlations for a trace" >}}

## Before you begin

> 

Reviewer: ```suggestion

To use trace correlations, you need: 

```

- Grafana 12 or later

> - A Tempo tracing data source configured in Grafana

Reviewer: ```suggestion
- A [Tempo data source](https://grafana.com/docs/grafana/<GRAFANA_VERSION>/datasources/tempo/configure-tempo-data-source/) configured in Grafana
```

- Admin access to configuration settings or provisioning files in Grafana

## Set up a trace correlation

1. Log in to Grafana with an admin account.

> 1. Go to **Configuration** → **Plugins & data** → **Correlations**.

Reviewer: ```suggestion
1. Go to **Configuration** > **Plugins & data** > **Correlations**.
```

1. Select **Add correlation** or **Add new**.

> 1. On step 1, provide a **label** for the correlation, and an optional **description**.

Reviewer: Are there actual named steps in the UI? If so, then any descriptions in the text should be bolded to indicate that they are UI elements. Otherwise, saying "step 1" can be confusing.

Reviewer: Okay, I checked the preview link you provided never mind the above. There are visual queues for the user to follow.

Author: No action needed then.

1. On step 2, configure the correlation **target**.

>    - Select the target type: **query** to link to another data source, or **external** for a custom URL.

Reviewer: ```suggestion
   - In the **Type** drop-down list, select **Query** to link to another data source, or **External** for a custom URL.
```

>    - For a query target, select the **target data source** and define the target query.

Reviewer: ```suggestion
   - For a query **Target**, select the target drop-down list and select the data source that should be queried when the link is clicked. Define the target query.
```

   - For an external target, enter the **external URL**.

   - For both query and external targets, you can use the following variables based on trace data. Object variables must be parsed into a value variable with a regular expression transformation.

   | Variable       | Type   | Description            |
   | -------------- | ------ | ---------------------- |
   | `traceId`      | String | Trace identifier       |
   | `spanID`       | String | Span identifier        |
   | `parentSpanID` | String | Parent span identifier |
   | `serviceName`  | String | Service name           |
   | `serviceTags`  | Object | Resource attributes    |
   | `tags`         | Object | Span attributes        |
   | `logs`         | Object | Trace events           |
   | `references`   | Object | Trace links            |

   {{< figure src="/media/docs/tempo/screenshot-grafana-trace-correlations-loki-step-2.png" max-width="900px" class="docs-image--no-shadow" alt="Setting up a correlation for a Loki target using trace variables" >}}

> 1. On step 3, configure the correlation **data source**:

Reviewer: ```suggestion
1. On step 3, configure the correlation data source:
```

   - Select your Tempo data source as **source**.

>    - Enter the trace data variable you use for the correlation in the **Result field**.

Reviewer: ```suggestion
   - Enter the trace data variable you use for the correlation in the **Results field**.
```

   - Optionally, add one or more **transformations** to parse the trace data into additional variables. You can use these variables to configure the correlation **target**.

   {{< figure src="/media/docs/tempo/screenshot-grafana-trace-correlations-loki-step-3.png" max-width="900px" class="docs-image--no-shadow" alt="Setting up a correlation for a Loki data source" >}}

1. Select **Save** to save the correlation.

## Verifying correlations in Explore

1. Open **Explore** and select your Tempo tracing source.

1. Run a query to load spans.

> 1. Hover over a the span links menu or open the span details to reveal the correlation link buttons.

Reviewer: ```suggestion
1. Hover over the span links menu or open the span details to reveal the correlation link buttons.
```

1. On step 2, configure the correlation target:

   - Select the target type **Query** and select your Loki data source as **Target**.

   - Define the Loki query, using `serviceName` and `traceID` as variables derived from the span data:

     ```
     {service_name="$serviceName"} | trace_id=`$traceID` |= ``
     ```

     {{< figure src="/media/docs/tempo/screenshot-grafana-trace-view-correlations-example-1-step-2.png" max-width="900px" class="docs-image--no-shadow" alt="Using correlations for a trace" >}}

1. On step 3, configure the correlation source:

   - Select your Tempo data source as **Source**.

   - Use `traceID` as **Results field**.

   - Add a new transformation to extract the `serviceName` from the span `serviceTags` using the regular expression:

     ```
     {(?=[^\}]*\bkey":"service.name")[^\}]*\bvalue":"(.*?)".*}
     ```

   {{< figure src="/media/docs/tempo/screenshot-grafana-trace-view-correlations-example-1-step-3.png" max-width="900px" class="docs-image--no-shadow" alt="Using correlations for a trace" >}}

1. Save the correlation.

> ### 2. Trace to custom URL

Reviewer: ```suggestion
### Example 2: Trace to custom URL

In this example, you configure trace corrections with a custom URL. 

```

1. On step 1, add a new correlation with the label **Open custom URL** and an optional description.

1. On step 2, configure the correlation target:

   - Select the target type **External**.

   - Define your target URL, using variables derived from the span data. In this example, we are using `serviceName` and `traceID`.

     ```
     https://my-server.example.com/service=$serviceName&trace=$traceID
     ```

1. On step 3, configure the correlation source:

   - Select your Tempo data source as **Source**.

   - Use `traceID` as **Results field**.

   - Add a new transformation to extract the `serviceName` from the span `serviceTags` using the regular expression:

     ```
     {(?=[^\}]*\bkey":"service.name")[^\}]*\bvalue":"(.*?)".*}
     ```

1. Save the correlation.

## Best practices

> - **Name clearly:** Use descriptive names indicating source and target (e.g., **Trace to errors in logs**).

Reviewer: ```suggestion
- **Name clearly:** Use descriptive names indicating source and target. For example:  **Trace to errors in logs**.
```

## What are trace correlations?

Trace correlations let you define rules that inject context-sensitive links into your trace spans. When viewing traces in Explore or the Traces panel, users can click these links to navigate directly to relevant queries or URLs. Correlations are similar but more flexible to the [trace to logs, metrics, and profiles links you can configure for the Tempo data source](/docs/grafana/<GRAFANA_VERSION>/datasources/tempo/configure-tempo-data-source).

{{< figure src="/media/docs/tempo/screenshot-trace-view-correlations.png" max-width="900px" class="docs-image--no-shadow" alt="Using correlations for a trace" >}}

## Before you begin

- Grafana 12 or later
- A Tempo tracing data source configured in Grafana
- Admin access to configuration settings or provisioning files in Grafana

## Set up a trace correlation

1. Log in to Grafana with an admin account.

1. Go to **Configuration** → **Plugins & data** → **Correlations**.

> 1. Select **Add new**.

Reviewer 2: The screen says "Add correlation"

<img width="553" alt="Screenshot 2025-04-23 at 11 29 37" src="https://github.com/user-attachments/assets/746b773a-9acb-42fc-9d57-20513910e15a" />

Author: Good catch! Since I've already had some - the button said "Add new" - will update
![Screenshot 2025-04-23 at 12 32 15 PM](https://github.com/user-attachments/assets/c72a305d-aab1-42d6-b3c6-d54858e6a122)

Reviewer 2: Oh, you're right! There's add new when there are already some correlations in

     ```

   {{< figure src="/media/docs/tempo/screenshot-grafana-trace-view-correlations-example-1-step-3.png" max-width="900px" class="docs-image--no-shadow" alt="Using correlations for a trace" >}}

1. Save the correlation.

### 2. Trace to custom URL

1. On step 1, add a new correlation with the label **Open custom URL** and an optional description.

1. On step 2, configure the correlation target:

   - Select the target type **External**.

   - Define your target URL, using variables derived from the span data. In this example, we are using `serviceName` and `traceID`.

     ```
     https://my-server.example.com/service=$serviceName&trace=$traceID
     ```

1. On step 3, configure the correlation source:

   - Select your Tempo data source as **Source**.

   - Use `traceID` as **Results field**.

   - Add a new transformation to extract the `serviceName` from the span `serviceTags` using the regular expression:

     ```
     {(?=[^\}]*\bkey":"service.name")[^\}]*\bvalue":"(.*?)".*}
     ```

1. Save the correlation.

## Best practices

- **Name clearly:** Use descriptive names indicating source and target (e.g., **Trace to errors in logs**).

> - **Limit scope**: For high-cardinality fields (like `traceID`), ensure your target system can handle frequent queries.

Reviewer 2: Correlations are processed for each row in frontend so cardinality shouldn't affect the performance

Author: Here I meant that that the target (e.g. Loki, etc.) is able to query efficiently whatever variables are used in the correlation query/URL.
But if you think it's ambiguous I'll remove this point.

- **Template wisely:** Use multiple `$variable` tokens if you need to inject more than one field.

> - **Test links:** After configuration, verify links on representative traces before rolling out broadly.

Reviewer 2: > (...) before rolling out broadly.

It suggests a bit that you can test the link before rolling out but in practice once it's saved it's available on the instance and there's no way to control the rollout

Author: Agreed, it makes sense to remove this point - I think (or just say... "don't forget to test it!", but that sounds silly)

---
description: Use Grafana correlations with Tempo traces
keywords:
  - grafana
  - tempo
  - guide
  - tracing
  - correlations
labels:
  products:
    - cloud
    - enterprise
    - oss
menuTitle: Trace correlations
title: Trace correlations
weight: 1000
---

# Trace correlations

You can use Grafana [correlations](/docs/grafana/<GRAFANA_VERSION>/administration/correlations/) to embed interactive correlation links in your trace view to jump from spans to related logs, metrics, profiles, or external systems. This guide explains how to configure and manage Trace correlations in Grafana.

## What are trace correlations?

> Trace correlations allow you to define rules that inject context-sensitive links into your trace spans. When viewing traces in Explore or the Traces panel, users can click these links to navigate directly to relevant queries or URLs. Correlations are similar but more flexible to the [trace to logs, metrics, and profiles links you can configure for the Tempo data source](/docs/grafana/<GRAFANA_VERSION>/datasources/tempo/configure-tempo-data-source).

Reviewer: ```suggestion
Trace correlations let you  define rules that inject context-sensitive links into your trace spans. When viewing traces in Explore or the Traces panel, users can click these links to navigate directly to relevant queries or URLs. Correlations are similar but more flexible to the [trace to logs, metrics, and profiles links you can configure for the Tempo data source](/docs/grafana/<GRAFANA_VERSION>/datasources/tempo/configure-tempo-data-source).
```

{{< figure src="/static/img/docs/tempo/screenshot-trace-view-correlations.png" max-width="900px" class="docs-image--no-shadow" alt="Using correlations for a trace" >}}

## Before you begin

> - Grafana 12 or later

Reviewer: Is there a specific version of Tempo that's required?

Author: I don't believe so, it's purely a Grafana feature

- A Tempo tracing data source configured in Grafana

> - Admin access to Grafana’s configuration settings or provisioning files

Reviewer: We try not to use product names as possessives. 
```suggestion
- Admin access to configuration settings or provisioning files in Grafana
```
